Mathlib/Algebra/Lie/Weights.lean

```diff
@@ -0,0 +1,514 @@
+      map_lie' := fun {x y} => by
+        ext m
+        simp only [rootSpaceWeightSpaceProductAux, LieSubmodule.coe_bracket,
+          LieSubalgebra.coe_bracket_of_module, lie_lie, LinearMap.coe_mk, AddHom.coe_mk,
+          Subtype.coe_mk, LieHom.lie_apply, LieSubmodule.coe_sub] }
+#align lie_algebra.root_space_weight_space_product LieAlgebra.rootSpaceWeightSpaceProduct
+
+@[simp]
+theorem coe_rootSpaceWeightSpaceProduct_tmul (χ₁ χ₂ χ₃ : H → R) (hχ : χ₁ + χ₂ = χ₃)
+    (x : rootSpace H χ₁) (m : weightSpace M χ₂) :
+    (rootSpaceWeightSpaceProduct R L H M χ₁ χ₂ χ₃ hχ (x ⊗ₜ m) : M) = ⁅(x : L), (m : M)⁆ := by
+  simp only [rootSpaceWeightSpaceProduct, rootSpaceWeightSpaceProductAux, coe_liftLie_eq_lift_coe,
+    AddHom.toFun_eq_coe, LinearMap.coe_toAddHom, lift_apply, LinearMap.coe_mk, AddHom.coe_mk,
+    Submodule.coe_mk]
+#align lie_algebra.coe_root_space_weight_space_product_tmul LieAlgebra.coe_rootSpaceWeightSpaceProduct_tmul
+
+/-- Given a nilpotent Lie subalgebra `H ⊆ L` together with `χ₁ χ₂ : H → R`, there is a natural
+`R`-bilinear product of root vectors, compatible with the actions of `H`. -/
+def rootSpaceProduct (χ₁ χ₂ χ₃ : H → R) (hχ : χ₁ + χ₂ = χ₃) :
+    rootSpace H χ₁ ⊗[R] rootSpace H χ₂ →ₗ⁅R,H⁆ rootSpace H χ₃ :=
+  rootSpaceWeightSpaceProduct R L H L χ₁ χ₂ χ₃ hχ
+#align lie_algebra.root_space_product LieAlgebra.rootSpaceProduct
+
+@[simp]
+theorem rootSpaceProduct_def : rootSpaceProduct R L H = rootSpaceWeightSpaceProduct R L H L := rfl
+#align lie_algebra.root_space_product_def LieAlgebra.rootSpaceProduct_def
+
+theorem rootSpaceProduct_tmul (χ₁ χ₂ χ₃ : H → R) (hχ : χ₁ + χ₂ = χ₃) (x : rootSpace H χ₁)
+    (y : rootSpace H χ₂) : (rootSpaceProduct R L H χ₁ χ₂ χ₃ hχ (x ⊗ₜ y) : L) = ⁅(x : L), (y : L)⁆ :=
+  by simp only [rootSpaceProduct_def, coe_rootSpaceWeightSpaceProduct_tmul]
+#align lie_algebra.root_space_product_tmul LieAlgebra.rootSpaceProduct_tmul
+
+/-- Given a nilpotent Lie subalgebra `H ⊆ L`, the root space of the zero map `0 : H → R` is a Lie
+subalgebra of `L`. -/
+def zeroRootSubalgebra : LieSubalgebra R L :=
+  { (rootSpace H 0 : Submodule R L) with
+    lie_mem' := fun {x y hx hy} => by
+      let xy : rootSpace H 0 ⊗[R] rootSpace H 0 := ⟨x, hx⟩ ⊗ₜ ⟨y, hy⟩
+      suffices (rootSpaceProduct R L H 0 0 0 (add_zero 0) xy : L) ∈ rootSpace H 0 by
+        rwa [rootSpaceProduct_tmul, Subtype.coe_mk, Subtype.coe_mk] at this
+      exact (rootSpaceProduct R L H 0 0 0 (add_zero 0) xy).property }
+#align lie_algebra.zero_root_subalgebra LieAlgebra.zeroRootSubalgebra
+
+@[simp]
+theorem coe_zeroRootSubalgebra : (zeroRootSubalgebra R L H : Submodule R L) = rootSpace H 0 := rfl
+#align lie_algebra.coe_zero_root_subalgebra LieAlgebra.coe_zeroRootSubalgebra
+
+theorem mem_zeroRootSubalgebra (x : L) :
+    x ∈ zeroRootSubalgebra R L H ↔ ∀ y : H, ∃ k : ℕ, (toEndomorphism R H L y ^ k) x = 0 := by
+  rw [zeroRootSubalgebra]
+  -- Porting note: added the following `change` otherwise the `simp` fails
```

Comment:
```suggestion
  -- Porting note: added the following `change` otherwise the `simp` fails
  -- See https://github.com/leanprover-community/mathlib4/issues/5026
```